### PR TITLE
nixos/appimage: init

### PR DESCRIPTION
**Pull request by jopejoe1**
## Description of changes
Adds an option to add and configure appimage-run as a system package and as binfmt executer.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/doc/manual/release-notes/rl-2405.section.md [hunk 1]
```diff
@@ -87,6 +87,8 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
 
 - [transfer-sh](https://github.com/dutchcoders/transfer.sh), a tool that supports easy and fast file sharing from the command-line. Available as [services.transfer-sh](#opt-services.transfer-sh.enable).
 
+- [AppImage](https://appimage.org/), a tool that supports running AppImages on nixos. Available as [programs.appimage](#opt-programs.appimage.enable).
```
Comment (lucasew) [inline nixos/doc/manual/release-notes/rl-2405.section.md:90]: `nixos` -> `NixOS`

File: nixos/modules/programs/appimage.nix [hunk 2]
```diff
@@ -0,0 +1,36 @@
+{ lib, config, pkgs, ... }:
+
+let
+  cfg = config.programs.appimage;
+  package = pkgs.appimage-run.override {extraPkgs = pkgs: cfg.extraPackages;};
+in
+
+{
+  options = {
+    programs.appimage = {
```
Comment (lucasew) [inline nixos/modules/programs/appimage.nix:10]: Suggestion: allow to override the package.

Basically `programs.appimage.package`.

Comment (jopejoe1) [reply nixos/modules/programs/appimage.nix:10]: Added, but not sure how useful that really is.

**Comment by jopejoe1** [comment]
@ofborg eval

File: nixos/doc/manual/release-notes/rl-2405.section.md [hunk 3]
```diff
@@ -87,6 +87,8 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
 
 - [transfer-sh](https://github.com/dutchcoders/transfer.sh), a tool that supports easy and fast file sharing from the command-line. Available as [services.transfer-sh](#opt-services.transfer-sh.enable).
 
+- [AppImage](https://appimage.org/), a tool that supports running AppImages on NixOS. Available as [programs.appimage](#opt-programs.appimage.enable).
```
Comment (Atemu) [inline nixos/doc/manual/release-notes/rl-2405.section.md:90]: This is a bit weird as it suggests this is somehow new and doesn't name the wrapper script (appimage-run). I'd rather announce the binfmt option; it's the star of the show I think.

**Review by Atemu — COMMENTED** [review]
So I was just typing down pretty much the exact same module and was ready to create a PR but it looks like you beat me to it :D

A couple improvements:

File: nixos/modules/programs/appimage.nix [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+      enable = lib.mkEnableOption (lib.mdDoc "appimage-run");
+      extraPackages = lib.mkOption {
+        type = lib.types.listOf lib.types.package;
+        default = [];
+        description = lib.mdDoc ''
+          List of packages to provide to appimage-run.
+        '';
+      };
+      binfmt = lib.mkEnableOption (lib.mdDoc "binfmt to directly execute AppImage's");
```
[the remark this answers is not in the document]
Comment (jopejoe1) [reply nixos/modules/programs/appimage.nix]: didn't realise lib.mdDoc was not necessary

File: nixos/doc/manual/release-notes/rl-2405.section.md (hunk 3 above)
Comment (jopejoe1) [reply nixos/doc/manual/release-notes/rl-2405.section.md]: Does this look better now?

```markdown
- binfmt option for AppImage-run to support running [AppImage](https://appimage.org/)'s seamlessly on NixOS.. Available as [programs.appimage.binfmt](#opt-programs.appimage.binfmt).
```

**Review by Atemu — APPROVED** [review]

**Pull request merged**